Added a "BeginInvokeOnMainThreadAsync" method to "IDeviceService"

rssllgrrtt opened the pull request:
﻿## Description of Change

Added an async version of "BeginInvokeOnMainThread as detailed here (https://github.com/PrismLibrary/Prism/issues/2043)

I've not written any tests, as I couldn't see any existing tests for that class. I'm happy to write some tests for that class if required, unless I've just missed existing ones.

### Bugs Fixed

- Provide links to bugs here

### API Changes

Added:

- Task IDeviceService.BeginInvokeOnMainThreadAsync(Action)
- Task DeviceService.BeginInvokeOnMainThreadAsync(Action)

### Behavioral Changes

The developer now has the ability to await Actions invoked on the main thread.

### PR Checklist

- [x ] Has tests (if omitted, state reason in description)
- [ x] Rebased on top of master at time of PR
- [ x] Changes adhere to coding standard

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/PrismLibrary/Prism?pullRequest=2103) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/PrismLibrary/Prism?pullRequest=2103) before we can accept your contribution.<br/><br/>:x: rssllgrrtt [sign now](https://cla.dotnetfoundation.org/PrismLibrary/Prism?pullRequest=2103)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/PrismLibrary/Prism?pullRequest=2103) it.</sub>

rssllgrrtt commented:
Hey. 

I'm really new to working on OS, but I'm keen to get started and help as much as I can. Please do let me know if I've not got anything right about this, I'm eager to learn.

Cheers!

dansiegel commented on src/Forms/Prism.Forms/Services/DeviceService.cs:
```diff
@@ -66,6 +67,30 @@ public void BeginInvokeOnMainThread(Action action)
+        /// Invokes an action (which can be awaited) on the device main UI thread.
+        /// </summary>
+        /// <param name="action">The Action to invoke</param>
+        public Task BeginInvokeOnMainThreadAsync(Action action)
+        {
+            var tcs = new TaskCompletionSource<bool>();
+            
+            Device.BeginInvokeOnMainThread(() =>
+            {
+                try
+                {
+                    action();
+                    tcs.TrySetResult(true);
+                }
+                catch (Exception ex)
+                {
+                    tcs.TrySetException(ex);
+                }
+            });
+            
+            return tcs.Task;
+        }
```
This is not something that we will be adding. Xamarin.Forms has an `InvokeOnMainThreadAsync` which we can map to, but we're not going to add an entirely new implementation.

rssllgrrtt replied:
Ok. Did I miss something then? Should I base this on the shell branch, as I’m assuming that’s built against a newer version of XF?

dansiegel replied:
go ahead and update to 4.5. You'll see the new method signatures on the Device class there

rssllgrrtt replied:
Great. I was just being a little wary, I didn’t know if there was a specific reason that it hadn’t been updated. I’ll get that sorted. Thanks for the clarification. 👍

dansiegel replied:
We don't update Xamarin.Forms just because there are updates available. That said Prism 8 will be updating as there are some changes that we will adjusting for

rssllgrrtt replied:
Makes complete sense. I’ll bear that in mind going forward. I’m in the U.K, so it’s bedtime for me now. I’ll get this sorted tomorrow. Thanks for your time.

rssllgrrtt replied:
OK. I've made the changes now.

I did notice a "Device.GetMainThreadSynchronizationContextAsync" method has been added to the XF device class too. Should support also be provided for this? If so, I'm happy to do that as well.

dansiegel replied:
anything that no longer maps properly should be removed, and any new additions can be added to the DeviceService

rssllgrrtt commented:
OK. I've made the changes now. 

I did notice a "Device.GetMainThreadSynchronizationContextAsync" method has been added to the XF device class too. Should support also be provided for this? If so, I'm happy to do that as well.

dansiegel commented on src/Forms/Prism.Forms/Prism.Forms.csproj:
```diff
@@ -12,7 +12,7 @@
 
   <ItemGroup>
-    <PackageReference Include="Xamarin.Forms" Version="3.6.0.344457" />
+    <PackageReference Include="Xamarin.Forms" Version="4.5.0.657" />
```
this is not compatible with MonoAndroid81... please change the TargetFrameworks to:

```xml
<TargetFrameworks>netstandard2.0;MonoAndroid90;MonoAndroid10.0</TargetFrameworks>
```

dansiegel replied:
I should add the Xamarin.Forms team released 4.6 today, go ahead and update to 4.6

rssllgrrtt replied:
No worries. I’ve been a bit busy with boring, real life stuff for the last few days. But I’ll be back to this over the weekend. Cheers!

dansiegel commented:
I should add here, that you haven't updated Xamarin.Forms on the E2E app... please update Xamarin.Forms there as well

rssllgrrtt commented:
I'll get all the changes made this coming weekend, cheers!

rssllgrrtt commented:
Hey.

In this commit I've updated to the latest XF, updated the TargetFrameworks and used the XF API documentation to go through the device class and everything that was missing.

The only thing I'm confused about, the E2E app. Do you mean this one? https://github.com/PrismLibrary/Prism-Samples-Forms/tree/master/sample-apps/PrismFormsGallery

I promise I'll stop asking as many questions when I'm comfortable with how everything is set up, I just want to make sure before I commit a bunch of stuff and waste everyone's time.

Cheers.

dansiegel commented on src/Forms/Prism.Forms/Services/IDeviceService.cs:
```diff
@@ -30,6 +43,70 @@ public interface IDeviceService
+        /// <param name="size">The named size for which to get the numeric size.</param>
+        /// <param name="targetElementType">The element type for which to calculate the numeric size.</param>
+        /// <returns>The named size for the font that is used on the element on the native platform.</returns>
+        double GetNamedSize(NamedSize size, Type targetElementType);
```
this is very view centric and needs to be removed.

rssllgrrtt replied:
Fixed in the latest commit.

rssllgrrtt commented on src/Forms/Prism.Forms/Services/DeviceService.cs:
```diff
@@ -66,6 +79,97 @@ public void BeginInvokeOnMainThread(Action action)
+        public double GetNamedSize(NamedSize size, Type targetElementType)
+        {
+            return Device.GetNamedSize(size, targetElementType);
+        }
```
Fixed in the latest commit.

dansiegel commented on tests/Forms/Prism.DryIoc.Forms.Tests/Prism.DryIoc.Forms.Tests.csproj:
```diff
@@ -10,7 +10,7 @@
     <PackageReference Include="Moq" Version="4.13.1" />
-    <PackageReference Include="Xamarin.Forms" Version="3.6.0.344457" />
+    <PackageReference Include="Xamarin.Forms" Version="4.6.0.726" />
     <PackageReference Include="Xamarin.Forms.Mocks" Version="3.5.0.3" />
```
The tests are currently failing for Prism.Forms. This is probably because Xamarin.Forms.Mocks will need to be updated as there is a binary break that required an update from 3.5.0.3 to 4.6.0.1... please be sure to update this

rssllgrrtt commented:
Awesome! Thanks for your patience with my first ever PR to a public repo. I learned a ton about Prism just reading through all the code. I promise I’ll ask less questions in future contributions. 😀

dansiegel commented:
we all have to start somewhere. Thanks for the PR
